subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,373 @@
+/*
+ * Copyright (c) 2019 Manivannan Sadhasivam <mani@kernel.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+#include <errno.h>
+#include <lorawan.h>
+#include <zephyr.h>
+
+#include <LoRaMac.h>
+
+#ifdef CONFIG_LORAMAC_REGION_AS923
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_AS923
+#elif CONFIG_LORAMAC_REGION_AU915
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_AU915
+#elif CONFIG_LORAMAC_REGION_CN470
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_CN470
+#elif CONFIG_LORAMAC_REGION_CN779
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_CN779
+#elif CONFIG_LORAMAC_REGION_EU433
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_EU433
+#elif CONFIG_LORAMAC_REGION_EU868
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_EU868
+#elif CONFIG_LORAMAC_REGION_KR920
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_KR920
+#elif CONFIG_LORAMAC_REGION_IN865
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_IN865
+#elif CONFIG_LORAMAC_REGION_US915
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_US915
+#elif CONFIG_LORAMAC_REGION_RU864
+	#define CONFIG_LORAWAN_REGION LORAMAC_REGION_RU864
+#elif
+	#error "Atleast one LoRaWAN region should be selected"
+#endif
+
+#define LOG_LEVEL CONFIG_LORAWAN_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(lorawan);
+
+K_SEM_DEFINE(lorawan_config_sem, 0, 1);
+K_SEM_DEFINE(lorawan_tx_sem, 0, 1);
+
+K_MUTEX_DEFINE(lorawan_join_mutex);
+K_MUTEX_DEFINE(lorawan_send_mutex);
+
+static volatile bool join_status = false;
+static volatile bool send_status = false;
+
+/* MAC status strings */
+const char *to_status_str[] = {
```

Comment:
Converting the status value to string, would be less error prone if you do the conversion directly in a function, so something like this:
```
const char *status2str(int status)
{
    switch (status) {
    case LORAMAC_STATUS_OK:
         return "OK";
....
}
```

Reply:
Okay